Rbac descendants

kbrock opened the pull request:
Larger goal is to use sql to determine valid ids instead of bring all resource ids back

1. remove the N+1 query that fetches every parent resource and every descendant resource
2. convert `match_via_descendants` option to just take a class (and simplify our logic)

/cc @gtanzillo I cut up into commits that are pretty tame. may want to go one by one.
/cc @matthewd thanks for your help so far here.
**UPDATED:** rewrite description to focus on `descendant_resource`

matthewd commented on app/models/rbac.rb:
```diff
@@ -331,68 +329,48 @@ def self.filtered(objects, options = {})
-  def self.find_descendants(descendant_klass, options)
-    search(options.merge(:class => descendant_klass, :results_format => :objects)).first
+      match.select { |m| m.kind_of?(klass) }
+    end.uniq
```
```
descendants.flat_map do |object|
  object.send(method_name)
end.grep(klass).uniq
```

(or use `select` if the `grep` is too obtuse -- but put it outside, so the `flat_map` can take care of arrayification for us)

kbrock replied:
yea, thanks - I want to move that `grep`/`select` into the original query, so others are not even brought back.

chessbyte commented:
LGTM :+1: 
@gtanzillo please review

kbrock commented:
- fixed rubocops.
- removed `pluck_ids` which was not yet necessary and made diff comparison more complicated.
- merged 2 commits that both changed `Rbac.find_via_descendant trim`

Hope this makes it easier to review

gtanzillo commented on app/models/rbac.rb:
```diff
@@ -331,22 +331,16 @@ def self.filtered(objects, options = {})
-    end
-    matches
+    MiqPreloader.preload(descendants, method_name)
+    descendants.flat_map { |object| object.send(method_name) }.grep(klass).uniq
```
Very cool!

gtanzillo commented:
:+1:  Looks good to me.
